`@no_type_check` support (#15122)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/suppressions/no-type-check.md

Reviewer: nit: we use snake_case rather than kebab-case for most of our other mdtest filenames (and the decorator is called `no_type_check` rather than `no-type-check`, so snake case would make more sense to me here even if that wasn't our convention)

# `@no_type_check`

> If a type checker supports the `no_type_check` decorator for functions, it should suppress all
> type errors for the def statement and its body including any nested functions or classes. It
> should also ignore all parameter and return type annotations and treat the function as if it were
> unannotated. [source](https://typing.readthedocs.io/en/latest/spec/directives.html#no-type-check)

## Error in the function body

```py
from typing import no_type_check

@no_type_check
def test() -> int:
    return a + 5
```

## Error in nested function

```py
from typing import no_type_check

@no_type_check
def test() -> int:
    def nested():
        return a + 5
```

## Error in nested class

```py
from typing import no_type_check

@no_type_check
def test() -> int:
    class Nested:
        def inner(self):
            return a + 5
```

## Error in preceding decorator

Don't suppress diagnostics for decorators appearing before the `no_type_check` decorator.

```py
from typing import no_type_check

@unknown_decorator  # error: [unresolved-reference]

> @no_type_check

Reviewer: Honestly, I would much rather we give this decorator the same semantics as any other decorator in the language has at runtime when it comes to the things the decorator applies to. We can have a debate over whether or not that's intuitive to people who aren't fully familiar with the semantics of decorators, but doing otherwise just feels really inconsistent to me. Decorators have very consistent semantics on this point in Python (the desugaring I gave in https://github.com/astral-sh/ruff/pull/15122#discussion_r1896765782), but we'd be saying we'd be treating this one decorator differently to any other decorator in the language (and, in fact, differently to how it works at runtime!). I'm honestly not sure how we'd justify the different semantics for this decorator in documentation or if we were responding to issue reports; it just doesn't make sense.

I think mypy and pyright got it wrong here, and we have a chance to do better.

> @no_type_check

Author: > I think mypy and pyright got it wrong here, and we have a chance to do better.

Maybe, but do we feel that strongly that it's worth introducing an incompatibility? What you propose is also not consistent with the specification (regardless of whatever reading you use). Overall I feel like consistency with other tools and the specification is more important than deviating because it could be an annoyance for users using multiple type checkers. It's also something that's very easy to change later (and probably so uncommon, that getting this perfect has very diminishing return)

> @no_type_check

Reviewer 2: I regret bringing this up 😆 

> considering that this most closely matches your reading of the spec

No, my reading of the spec wording is that it would suggest suppressing errors in all decorators (they are part of the def statement, because they modify it and they can't stand apart from it), which is the opposite of what mypy and pyright do.

I don't think there is a consensus here that we'd be deviating from, I just think it's an edge-case behavior that nobody ever thought about very carefully, either in the existing type checkers or in writing the spec.

I'm also not concerned about this being an adoption barrier or compatibility problem, since it would only cause us to suppress a few additional type errors in an unusual case; worst case someone might have to remove a `type: ignore`? I think there will be many more serious problems for anyone trying to check a codebase on multiple different type checkers; I don't think people do this, because it's too painful. Compatibility between type checkers in practice is all about libraries being able to specify public interfaces that can work for users of multiple type checkers; this issue doesn't affect that.

If this ever came up in discussion in the typing council, I would argue for specifying the "suppress errors in later decorators" behavior, and I suspect there would be support for that position, because it's the most sensible one given Python semantics.

> @no_type_check

Reviewer: > I think there will be many more serious problems for anyone trying to check a codebase on multiple different type checkers; I don't think people do this, because it's too painful.

It's pretty common for library authors to do this, because library authors often want users of their library to have a good experience whether their users use pyright or mypy. And it's also pretty common for people to use VSCode or pycharm as their IDE locally (which usually means using pyright or pycharm in their editor, even though there _are_ mypy plugins for both) but run mypy in CI. So I don't think it's true that "people don't do this".

But I agree that people find it painful. I also agree that this is a very minor incompatibility relative to the many other incompatibilities that already exist between mypy/pyright/pycharm, and that are likely to exist between mypy/red-knot, pyright/red-knot and pycharm/red-knot. And I agree that people are very unlikely to come across this incompatibility in practice, and that there are easy workarounds for people who do come across it.

> @no_type_check

Author: I agree that there are very few users that will be impacted by this change, and if so, it's most likely only that a violation now has been fixed (which might raise an `unused-type-ignore` warning). However, my concern is that every deviation from mypy and pyright increases the migration cost to Red Knot. Is it because they have to change their mental model or because there are more, fewer, or different violations. That's why I think we should pick our battles carefully because large code bases are likely to exercise many deviations, and every one of those will require manual changes when migrating to Red Knot (or using Red Knot side by side with another type checker). 

My other concern is simply that I think this approach is the least spec-compliant because you could read it as either including or excluding all decorators. But excluding only some seems to be the most far-fetched interpretation. 

That's why I still think leaving it just as is for now is the best and we can easily change it later (it only requires moving a single line of code). However, I also get the impression that you two feel somewhat strongly about this, and that's why I don't object to changing it (just tell me, no need for an explanation ;)).

> @no_type_check

Reviewer: I honestly think it's a mistake to read the language of the spec too closely in situations like this. The aim is that the spec will eventually be precise in its language in all areas. But the spec is still very new, and still a work in progress in many places. I think there are many places in the spec where the language the spec uses could plausibly be understood to imply behaviours in certain edge cases which were never intended by authors of the spec.

I agree that compatibility with other type checkers is really important, so I really appreciate you arguing strongly for this! However, I still feel quite strongly that the order-dependent behaviour (where `@no_type_check` applies its magic to all decorators below `@no_type_check`, and the class, but no decorators above it) is the better one. If one thing is important for a Python type checker, it's for the tool to have a consistent and understandable model of Python's semantics with as few special cases as possible. So I _really_ dislike the idea of carving out a special case for this decorator so that it works differently to all other decorators in Python. In this case, I don't think that's outweighed by the compatibility costs, personally :-)

def test() -> int:
    return a + 5
```

## Error in following decorator

Suppress diagnostics for decorators appearing after the `no_type_check` decorator.

```py
from typing import no_type_check

@no_type_check
@unknown_decorator
def test() -> int:
    return a + 5
```

## Error in default value

```py
from typing import no_type_check

@no_type_check
def test(a: int = "test"):
    return x + 5
```

## Error in return value position

```py
from typing import no_type_check

@no_type_check
def test() -> Undefined:
    return x + 5
```

## `no_type_check` on classes isn't supported

Red Knot does not support `no_type_check` annotations on classes currently. The behaviour of
`no_type_check` when applied to classes is
[not fully specified currently](https://typing.readthedocs.io/en/latest/spec/directives.html#no-type-check),
and applying the decorator to classes is not supported by Pyright and possibly other type checkers.

> A future improvement might be to emit a diagnostic if a `no_type_check` annotation is applied to a

Author: Eh, but your suggested wording (thanks!) on line 89 used annotations. That's why I used it here.

> A future improvement might be to emit a diagnostic if a `no_type_check` annotation is applied to a

Reviewer: Oh, great point, the use of "annotation" on that line is also incorrect, and should be changed :-)

I didn't notice that you used the word "annotation" in that sentence when I made my suggestion in https://github.com/astral-sh/ruff/pull/15122#discussion_r1896836658; I guess I was concentrating on other things at the time. Sorry!

class.

```py
from typing import no_type_check

@no_type_check
class Test:
    def test(self):
        return a + 5  # error: [unresolved-reference]
```

## `type: ignore` comments in `@no_type_check` blocks

```py
from typing import no_type_check

@no_type_check
def test():

>     # error: [unused-ignore-comment]

Reviewer: could we assert the error message here, so that it's obvious in the future what's changing if we decide to improve the diagnostic in the future?

```suggestion
    # error: [unused-ignore-comment] "Unused `knot: ignore` directive: 'unresolved-reference'"
```

# `@no_type_check`

> If a type checker supports the `no_type_check` decorator for functions, it should suppress all type errors for the def statement and its body including any nested functions or classes. It should also ignore all parameter and return type annotations and treat the function as if it were unannotated.
> [source](https://typing.readthedocs.io/en/latest/spec/directives.html#no-type-check)

## Error in the function body

```py
from typing import no_type_check

@no_type_check
def test() -> int: 
    return a + 5
```

## Error in nested function

```py
from typing import no_type_check

@no_type_check
def test() -> int:
    def nested(): 
        return a + 5
```

## Error in nested class

```py
from typing import no_type_check

@no_type_check
def test() -> int:
    class Nested:
        def inner(self):
            return a + 5
```

## Error in decorator

> Both MyPy and Pyright flag the `unknown_decorator` but we don't. 

Author: Thanks for the extra input. I intentionally left this open for now because the approach on how to implement this probably depends on how we handle those diagnostics :) I somewhat agree but then again, `no_type_check` disables type checking for the entire function... which I would expect includes all decorators.

> Both MyPy and Pyright flag the `unknown_decorator` but we don't. 

Author: WIth later: I mainly mean, once we figured out how to implement this properly 😆

```py
from typing import no_type_check

@unknown_decorator
@no_type_check
def test() -> int:
    return a + 5
```

## Error in default value

```py
from typing import no_type_check

@no_type_check
def test(a: int = "test"): 
    return x + 5
```

## Error in return value position

```py
from typing import no_type_check

@no_type_check
def test() -> Undefined: 
    return x + 5
```

## `no_type_check` on classes isn't supported

> Similar to Pyright, Red Knot does not support `no_type_check` annotations on classes.

Author: sure: It's mainly that the semantics are unclear according to [this discussion](https://discuss.python.org/t/no-type-check-decorator/43119) and the typing spec explicitly leaves it open for type checkers not to support classes.

> Similar to Pyright, Red Knot does not support `no_type_check` annotations on classes.

Reviewer: Ah, I forgot that mypy also didn't support decorating classes with `@no_type_check` (and in fact, no other type checker does currently). Your wording here made me think that we were specifically copying pyright's implementation decisions here, but that's obviously not the case. Could you maybe reword this to something like this?

```suggestion
Red Knot does not support `no_type_check` annotations on classes currently. The behaviour of `no_type_check` when applied to classes is [not fully specified currently](https://typing.readthedocs.io/en/latest/spec/directives.html#no-type-check), and applying the decorator to classes is not supported by any other type checkers currently.
```

We could even consider emitting a diagnostic if a class is decorated with `@no_type_check`, since it clearly won't have the effect the user wants.

> Similar to Pyright, Red Knot does not support `no_type_check` annotations on classes.

Author: > We could even consider emitting a diagnostic if a class is decorated with @no_type_check, since it clearly won't have the effect the user wants.

Possibly, but not as part of this PR ;)

> Similar to Pyright, Red Knot does not support `no_type_check` annotations on classes.

Reviewer: > Possibly, but not as part of this PR ;)

Is it worth adding a TODO comment somewhere as part of this PR? ;)

```py
from typing import no_type_check

@no_type_check
class Test:
    def test(self):
        return a + 5  # error: [unresolved-reference]
```

## `type: ignore` comments in `@no_type_check` blocks

```py
from typing import no_type_check

@no_type_check
def test():

>     #  error: [unused-ignore-comment]

Reviewer: ```suggestion
    # error: [unused-ignore-comment]
```

Also, I wonder if it would be nice to have an additional explanation included in the error message here, such as:

```suggestion
    # error: [unused-ignore-comment] "Unused ignore comment (all errors in `test` are already ignored due because it is decorated with `@no_type_check`)"
```

Author: > Also, I wonder if it would be nice to have an additional explanation included in the error message here, such as:

Possibly, but it requires some extra metadata tracking for suppressions or at least post-processing in the `unused-ignore-comment` rule. I don't think it's a priority right now (we already go beyond by flagging unused ignore comments
